File: python/python/lancedb/secrets.py
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright The LanceDB Authors

"""Named Secrets, and the bindings that deliver them to Functions.

A Secret is a database-scoped named credential. Nothing in this module holds a
value: :class:`EnvVarSecret` names one and says which environment variable it
should arrive in, and the value is resolved by the remote service when a
Function bound to it runs. No API returns a stored credential, by construction
rather than by policy -- there is no code path that could.
"""

from __future__ import annotations

import re

# The same characters LanceDB already admits in a namespace or table name, and
# no positional rule on top of them: a segment may begin with `_`, `-` or `.`
# today, so anything narrower would put Secrets out of reach inside namespaces
# that already exist. Matches the service, which admits the same set.
_SECRET_NAME = re.compile(r"^[A-Za-z0-9_.-]{1,255}$")
_ENV_VARIABLE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


def validate_secret_name(name: str) -> str:
    """Check a Secret name locally and return it unchanged."""
    if not isinstance(name, str):
        raise TypeError(f"Secret name must be a string, not {type(name).__name__}")
    if not _SECRET_NAME.fullmatch(name):
        raise ValueError(f"invalid Secret name: {name!r}")
    return name


def validate_env_variable(name: str) -> str:
    """Check an environment variable name locally and return it unchanged."""
    if not isinstance(name, str):
        raise TypeError(
            f"environment variable name must be a string, not {type(name).__name__}"
        )
    if not _ENV_VARIABLE.fullmatch(name):
        raise ValueError(f"invalid environment variable name: {name!r}")
    return name


class EnvVarSecret:
    """A Secret bound to the environment variable a Function's library reads.

    Pass these in the ``secrets`` sequence of
    [DBConnection.create_function][lancedb.db.DBConnection.create_function]. The
    Function's source is unchanged by binding: it reads ``OPENAI_API_KEY`` the
    way it always did, and the binding is what puts a value there.

    This is a local value. Constructing it contacts no server, so it always
    succeeds and says nothing about whether the Secret exists; that is checked
    at registration, where a mistyped Secret name surfaces as a clear "does not
    exist" naming both the Secret and the variable bound to it. A mistyped
    *variable* name cannot be caught anywhere -- nothing knows which variables a
    Function reads -- so it surfaces on the first rows instead.

    The type exists so a credential cannot be passed by accident. A bare string
    in the same position is a plausible-looking mistake with the opposite
    meaning, and it reads identically in a diff.

    Parameters
    ----------
    secret : str
        The Secret's database-scoped name.
    env_variable : str
        The environment variable the Function reads it from.

    Examples
    --------
    >>> from lancedb import EnvVarSecret
    >>> binding = EnvVarSecret(secret="openai-prod", env_variable="OPENAI_API_KEY")
    >>> binding.secret, binding.env_variable
    ('openai-prod', 'OPENAI_API_KEY')
    """

    __slots__ = ("_secret", "_env_variable")

    def __init__(self, secret: str, env_variable: str):
        self._secret = validate_secret_name(secret)
        self._env_variable = validate_env_variable(env_variable)

    @property
    def secret(self) -> str:
        """The Secret's database-scoped name."""
        return self._secret

    @property
    def env_variable(self) -> str:
        """The environment variable the value is delivered in."""
        return self._env_variable

    def __repr__(self) -> str:
        return (
            f"EnvVarSecret(secret={self._secret!r}, "
            f"env_variable={self._env_variable!r})"
        )

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, EnvVarSecret)
            and other._secret == self._secret
            and other._env_variable == self._env_variable
        )

    def __hash__(self) -> int:
        return hash((EnvVarSecret, self._secret, self._env_variable))


class SecretInfo:
    """What a database records about a Secret. Never its value.

    Returned by
    [DBConnection.describe_secret][lancedb.db.DBConnection.describe_secret].
    """

    __slots__ = ("_name", "_created_at_millis", "_updated_at_millis")

    def __init__(self, name: str, created_at_millis: int, updated_at_millis: int):
        self._name = name
        self._created_at_millis = created_at_millis
        self._updated_at_millis = updated_at_millis

    @property
    def name(self) -> str:
        """The Secret's database-scoped name."""
        return self._name

    @property
    def created_at_millis(self) -> int:
        """When the Secret was created, in milliseconds since the Unix epoch."""
        return self._created_at_millis

    @property
    def updated_at_millis(self) -> int:
        """When the Secret's value was last rotated, in epoch milliseconds.

        The only observable that a rotation landed: no API returns a credential,
        so a caller confirms ``alter_secret`` took effect by watching this move.
        """
        return self._updated_at_millis

    @classmethod
    def from_json(cls, value: dict) -> "SecretInfo":
        return cls(
            name=value["name"],
            created_at_millis=value["created_at_millis"],
            updated_at_millis=value["updated_at_millis"],
        )

    def __repr__(self) -> str:
        return (
            f"SecretInfo(name={self._name!r}, "
            f"created_at_millis={self._created_at_millis!r}, "
            f"updated_at_millis={self._updated_at_millis!r})"
        )

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, SecretInfo)
            and other._name == self._name
            and other._created_at_millis == self._created_at_millis
            and other._updated_at_millis == self._updated_at_millis
        )


__all__ = [
    "EnvVarSecret",
    "SecretInfo",
    "validate_env_variable",
    "validate_secret_name",
]

File: rust/lancedb/src/secrets.rs
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright The LanceDB Authors

//! Named Secrets: database-scoped credentials a Function binds by name.
//!
//! Nothing here holds a credential. The verbs live on
//! [`crate::connection::Connection`], and none of them returns a value -- by
//! construction rather than by policy, so there is no code path that could.
//!
//! What a Secret is, how one is named, and how a Function binds one all live
//! here; [`crate::function`] holds the bindings a FunctionVersion records, the
//! way Sophon's Secret catalog and Function catalog divide the same two.

use serde::de;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_json::Value;

/// What a database records about a Secret. Never its value.
///
/// Returned by [`crate::connection::Connection::describe_secret`]. There is no
/// field for the credential and no method that could produce one.
#[derive(Debug, Clone, PartialEq, Eq, serde::Deserialize)]
pub struct SecretInfo {
    /// The Secret's database-scoped name.
    pub name: String,
    /// When the Secret was created, in milliseconds since the Unix epoch.
    pub created_at_millis: i64,
    /// When the Secret's value was last rotated, in milliseconds since the Unix
    /// epoch.
    ///
    /// This is the only observable that a rotation landed: no API returns a
    /// credential, so a caller confirms `alter_secret` took effect by watching
    /// this move.
    pub updated_at_millis: i64,
}

/// Where a Secret lives, carried as its parts rather than as one string.
///
/// Nothing here is parsed, so nothing can parse two ways. A joined id would
/// instead need a delimiter excluded from every name and segment, agreed on by
/// both sides.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct SecretReference {
    pub name: String,
    /// The namespace holding the Secret. Empty is the root, and is absent from
    /// the wire rather than sent empty: a binding states a namespace only when
    /// it has one.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub namespace_path: Vec<String>,
}

impl SecretReference {
    /// A Secret in the root namespace.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            namespace_path: Vec::new(),
        }
    }

    /// A Secret in `namespace_path`.
    pub fn in_namespace(name: impl Into<String>, namespace_path: Vec<String>) -> Self {
        Self {
            name: name.into(),
            namespace_path,
        }
    }
}

/// How a Secret reaches the Function that binds it.
///
/// One list rather than a field per delivery mode, so a mode added later is a
/// variant and the per-Function rules -- how many Secrets a Function may bind,
/// which ones it needs -- stay answerable from one place.
///
/// A mode this client does not know decodes rather than failing the whole
/// FunctionVersion, as [`PythonRuntimeSpec`] does for an unknown runtime. That
/// takes both halves: [`SecretBinding::Unrecognized`] gives the wire somewhere
/// to land, and `#[non_exhaustive]` denies callers an exhaustive match, so a
/// later mode arrives as a case they already had to handle. Its payload is
/// dropped -- the client does not proxy catalog values.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
#[non_exhaustive]
pub enum SecretBinding {
    /// Delivered as an environment variable, which the UDF's library already
    /// reads. The variable is the delivery target; the Secret is what fills it.
    Env {
        variable: String,
        /// Named `secret_ref` rather than `secret` because a Job payload is
        /// scanned server-side for credential-shaped keys, and a key called
        /// `secret` trips that guard whatever it actually holds.
        secret_ref: SecretReference,
    },
    /// A binding kind introduced by a newer server.
    Unrecognized { kind: String },
}

impl SecretBinding {
    /// The wire discriminator reported by Sophon.
    pub fn kind(&self) -> &str {
        match self {
            Self::Env { .. } => "env",
            Self::Unrecognized { kind } => kind,
        }
    }

    /// The environment variable this binding fills, or `None` for a kind that
    /// does not deliver through one.
    pub fn variable(&self) -> Option<&str> {
        match self {
            Self::Env { variable, .. } => Some(variable),
            Self::Unrecognized { .. } => None,
        }
    }

    /// The Secret bound, or `None` for a kind this client cannot read.
    pub fn secret(&self) -> Option<&SecretReference> {
        match self {
            Self::Env { secret_ref, .. } => Some(secret_ref),
            Self::Unrecognized { .. } => None,
        }
    }
}

#[derive(Deserialize)]
struct EnvSecretBindingWire {
    variable: String,
    secret_ref: SecretReference,
}

impl<'de> Deserialize<'de> for SecretBinding {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
        let value = Value::deserialize(deserializer)?;
        let kind = value
            .get("kind")
            .ok_or_else(|| de::Error::missing_field("kind"))?
            .as_str()
            .ok_or_else(|| de::Error::custom("secret binding kind must be a string"))?
            .to_string();
        match kind.as_str() {
            "env" => {
                let wire: EnvSecretBindingWire =
                    serde_json::from_value(value).map_err(de::Error::custom)?;
                Ok(Self::Env {
                    variable: wire.variable,
                    secret_ref: wire.secret_ref,
                })
            }
            _ => Ok(Self::Unrecognized { kind }),
        }
    }
}

impl Serialize for SecretBinding {
    fn serialize<S: Serializer>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error> {
        #[derive(Serialize)]
        struct EnvBindingRef<'a> {
            kind: &'static str,
            variable: &'a str,
            secret_ref: &'a SecretReference,
        }

        #[derive(Serialize)]
        struct UnrecognizedBindingRef<'a> {
            kind: &'a str,
        }

        match self {
            Self::Env {
                variable,
                secret_ref,
            } => EnvBindingRef {
                kind: "env",
                variable,
                secret_ref,
            }
            .serialize(serializer),
            Self::Unrecognized { kind } => UnrecognizedBindingRef { kind }.serialize(serializer),
        }
    }
}
